### PR TITLE
Azure file system library interface

### LINKED ISSUE #46: Guidance on custom filesystem
The past guide for this was here https://www.tensorflow.org/guide/extend/filesystem and I'm looking at porting the [azure blob storage file system](https://github.com/tensorflow/tensorflow/pull/21649) to this repo. 

I was wanting some guidance on what might have changed? I found the igfs registration here https://github.com/tensorflow/io/blob/master/tensorflow_io/ignite/ops/igfs_ops.cc and what seems to be the build target here https://github.com/tensorflow/io/blob/master/tensorflow_io/ignite/BUILD#L5-L43 for this file system. Are these two components what is likely from my point of view?

### PR DESCRIPTION
**Pull request by damienpontifex**
- Use scheme az://
- c++ tests
- python tests

There's a lot here and very open to discussion on how it's laid out in the tensorflow_io structure, how a user would import the az file system library components etc etc.

I have tests run from bazel for the c++ interface and now in the python interface. Python tests are having some issues so I've marked this as draft for now as I'd like to resolve them.

Still also to do is integrate this with the CI build which seems to be travis. I can run the unit tests against a docker container which emulates azure blob storage. Will have to add this to the CI setup for completeness there as well

Would resolve #46

### CODE REVIEW

**Comment by yongtang** [comment]
@damienpontifex Thanks for the PR! I tried to build on my local linux machine. It may require some small changes. The following diff may work:
```diff
diff --git a/WORKSPACE b/WORKSPACE
index dea2b04..0b24a1b 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -201,6 +201,15 @@ http_archive(
     ],
 )
 
+http_archive(
+    name = "com_google_googletest",
+    strip_prefix = "googletest-release-1.8.1",
+    sha256 = "9bf1fe5182a604b4135edc1a425ae356c9ad15e9b23f9f12a02e80184c3a249c",
+    urls = [
+        "https://github.com/google/googletest/archive/release-1.8.1.tar.gz",
+    ],
+)
+
 http_archive(
     name = "com_github_azure_azure_storage_cpplite",
     build_file = "//third_party:azure.BUILD",
diff --git a/tensorflow_io/azure/BUILD b/tensorflow_io/azure/BUILD
index fc2615b..91b6631 100644
--- a/tensorflow_io/azure/BUILD
+++ b/tensorflow_io/azure/BUILD
@@ -137,5 +137,7 @@ cc_binary(
     ],
     deps = [
         ":azfs",
+        "@com_google_googletest//:gtest",
+        "@com_google_googletest//:gtest_main"
     ],
 )
diff --git a/tensorflow_io/azure/azfs_ops.py b/tensorflow_io/azure/azfs_ops.py
index bbbc604..c03aa4c 100644
--- a/tensorflow_io/azure/azfs_ops.py
+++ b/tensorflow_io/azure/azfs_ops.py
@@ -21,9 +21,5 @@ from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
 
-from tensorflow.python.framework import load_library
-from tensorflow.python.platform import resource_loader
-
-file_system_library = resource_loader.get_path_to_datafile("_azfs.so")
-load_library.load_library(file_system_library)
-# load_library.load_file_system_library(file_system_library)
\ No newline at end of file
+from tensorflow_io import _load_library
+_load_library("_azfs.so")
diff --git a/third_party/azure.BUILD b/third_party/azure.BUILD
index 0a4a59b..957c363 100644
--- a/third_party/azure.BUILD
+++ b/third_party/azure.BUILD
@@ -112,7 +112,6 @@ cc_library(
     linkopts = select({
         "@bazel_tools//src/conditions:darwin": [],
         "//conditions:default": [
-            "-luuid",
         ]
     }),
     copts = ["-std=c++11"],
diff --git a/third_party/uuid.BUILD b/third_party/uuid.BUILD
index 51cd21a..870919e 100644
--- a/third_party/uuid.BUILD
+++ b/third_party/uuid.BUILD
@@ -34,6 +34,7 @@ cc_library(
   hdrs = [
       "libuuid/src/uuid.h"
   ],
+  copts = ["--std=c99", "-D_XOPEN_SOURCE=700"],
   strip_include_prefix = "libuuid/src",
   include_prefix = "uuid",
   defines = select({
```

**Comment by damienpontifex** [comment]
Thanks @yongtang I took a few parts of your diff and found some build targets to clean up as well. I also need Azure/azure-storage-cpplite#26 for build to work again on macOS with the azure storage cpp library.
With this change and the upcoming external PR, I built locally on macOS and in a linux docker container if you could try again?

**Comment by yongtang** [comment]
@damienpontifex Looks like we are almost there! The python 3.7 build has some errors. I think this is because `"-D_GLIBCXX_USE_CXX11_ABI=0",` was passed (in `tensorflow_io/azure/BUILD`).

Since `./configure.sh` automatically detect the ABI flag and pass it to bazel, this explicit passing of `D_GLIBCXX_USE_CXX11_ABI=0 ` is not needed anymore.

Can you remove this line in `tensorflow_io/azure/BUILD`?

**Comment by damienpontifex** [comment]
@yongtang with the linting failure: method overrides such as `setUp` in the python tests, is there way to ignore that as it's an override method and so can't conform to our naming linting pattern?

**Comment by yongtang** [comment]
@damienpontifex I typically use the following to suppress the lint error:
```
def setUp(self): # pylint: disable=invalid-name
```

**Comment by yongtang** [comment]
Also, we lint bazel files as well. If you have docker installed, you can run the following command (quite long but works) which automatically buildifier all BUILD files:
```
docker run -i -t --rm -v $PWD:/v -w /v --net=host golang:1.12 bash -x -e -c 'go get github.com/bazelbuild/buildtools/buildifier && buildifier  $(find . -type f \( -name WORKSPACE -or -name BUILD -or -name *.BUILD \))'
```

**Comment by damienpontifex** [comment]
@yongtang any clang-format or c++ linting?

File: tensorflow_io/azure/azfs_test.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    rows = 100
+    # Write data.
+    with gfile.Open(file_name, 'w') as w:
+      for i in range(rows):
+        w.write("This is row {}\n".format(i))
+
+    # Read data.
+    with gfile.Open(file_name, 'r') as r:
+      for i, line in enumerate(r):
+        self.assertEqual("This is row {}\n".format(i), line)
```
Comment (damienpontifex) [inline tensorflow_io/azure/azfs_test.py]: @yongtang any idea why this test might be failing? It writes the file as expected. When I download that file from Azure it's ascii format. When it does the read, it complains about not being able to decode 
`UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbb in position 8: invalid start byte`

Comment (yongtang) [reply tensorflow_io/azure/azfs_test.py]: @damienpontifex I can give it a try if there is a sample public azfs I could access (don't have azure account).

On a side note, maybe you could give `with gfile.Open(file_name, 'rb')` a try?

**Comment by yongtang** [comment]
@damienpontifex For C++ linting, initially clang-format was used for tensorflow repo. However, clang-format will generate different result depending on the version. If a developer use clang-format 3.5 and lint on Travis CI use 3.8 then they could be different. For that reason tensorflow repo stopped using clang-format for linting.

In tensorflow-io I haven't found a easy way to allow contributors to get a reliable result from a clang-format so we haven't enabled c++ linting yet.

**Comment by yongtang** [comment]
@damienpontifex I tried locally on Linux, it looks like the failed Linux build on Travis CI could be fixed with a later version of the commit from azure_storage_cpplite:

(from `21f5d6f` to `54bb4cf`)

```diff
diff --git a/WORKSPACE b/WORKSPACE
index 9a6b9e8..0efeac7 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -204,11 +204,11 @@ http_archive(
 http_archive(
     name = "com_github_azure_azure_storage_cpplite",
     build_file = "//third_party:azure.BUILD",
-    sha256 = "617b1762888a5024d27a436aa88e74f7f0bdd11d2a5898074f82db09c2716f61",
-    strip_prefix = "azure-storage-cpplite-21f5d6f3a6ef433b70026f25385925c4aba1df29",
+    sha256 = "57117f0231e0629d025d9803895186183942894d028b122b355c4d6a24f7c69b",
+    strip_prefix = "azure-storage-cpplite-54bb4cfb0444f65fd9b0f54792237be43a31788a",
     urls = [
-        "https://github.com/Azure/azure-storage-cpplite/archive/21f5d6f3a6ef433b70026f25385925c4aba1df29.zip",
-        "https://mirror.bazel.build/github.com/Azure/azure-storage-cpplite/archive/21f5d6f3a6ef433b70026f25385925c4aba1df29.zip",
+        "https://github.com/Azure/azure-storage-cpplite/archive/54bb4cfb0444f65fd9b0f54792237be43a31788a.zip",
+        "https://mirror.bazel.build/github.com/Azure/azure-storage-cpplite/archive/54bb4cfb0444f65fd9b0f54792237be43a31788a.zip",
     ],
 )
```

**Comment by damienpontifex** [comment]
@yongtang yes it can 😄 ironically that then breaks the macOS build. I have a PR open in that repo for the follow up mac fix. I was waiting on that to merge and then update the dependency. I could point it to my fork with the fix in it if we want to get things rolling while we wait and then update to offical azure repo later

**Comment by yongtang** [comment]
Thanks @damienpontifex, saw your PR  in azure-storage-cpplite 👍 

Bazel actually allows a patch to be applied (with `patch_args` and `patches` keywords specified in repo rule) while in WORKSPACE. With the following diff I think we could just make the build working (and update the WORKSPACE again once your PR in  azure-storage-cpplite is merged):
```diff
diff --git a/WORKSPACE b/WORKSPACE
index 9a6b9e8..a75843c 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -204,11 +204,15 @@ http_archive(
 http_archive(
     name = "com_github_azure_azure_storage_cpplite",
     build_file = "//third_party:azure.BUILD",
-    sha256 = "617b1762888a5024d27a436aa88e74f7f0bdd11d2a5898074f82db09c2716f61",
-    strip_prefix = "azure-storage-cpplite-21f5d6f3a6ef433b70026f25385925c4aba1df29",
+    patch_args = ["-p1"],
+    patches = [
+        "//third_party:azure.patch",
+    ],
+    sha256 = "57117f0231e0629d025d9803895186183942894d028b122b355c4d6a24f7c69b",
+    strip_prefix = "azure-storage-cpplite-54bb4cfb0444f65fd9b0f54792237be43a31788a",
     urls = [
-        "https://github.com/Azure/azure-storage-cpplite/archive/21f5d6f3a6ef433b70026f25385925c4aba1df29.zip",
-        "https://mirror.bazel.build/github.com/Azure/azure-storage-cpplite/archive/21f5d6f3a6ef433b70026f25385925c4aba1df29.zip",
+        "https://github.com/Azure/azure-storage-cpplite/archive/54bb4cfb0444f65fd9b0f54792237be43a31788a.zip",
+        "https://mirror.bazel.build/github.com/Azure/azure-storage-cpplite/archive/54bb4cfb0444f65fd9b0f54792237be43a31788a.zip",
     ],
 )
 
diff --git a/third_party/azure.patch b/third_party/azure.patch
new file mode 100644
index 0000000..fd46a0f
--- /dev/null
+++ b/third_party/azure.patch
@@ -0,0 +1,11 @@
+diff -Naur a/include/compare.h b/include/compare.h
+--- a/include/compare.h        2019-06-23 22:12:03.208481680 +0000
++++ b/include/compare.h        2019-06-23 22:12:53.555100409 +0000
+@@ -3,6 +3,7 @@
+ #ifdef __linux__
+ #include <string.h>
+ #else
++#include <cstring>
+ #include <string>
+ #endif
+ 
```

**Comment by damienpontifex** [comment]
@yongtang all passing now 🍾 
I should note, I've been running tests against local emulator but not with TensorFlow projects yet. I wanted to get a first version submitted so I can iterate on it vs waiting any longer.

**Comment by yongtang** [comment]
Thanks @damienpontifex 👍 

We have emulators setup through:
https://github.com/tensorflow/io/blob/master/.travis.yml#L12-L23

Wondering if you could share how you setup the emulator for Azure? I think having maybe one test setup would be good enough for now and we cal always refine later.

I can help set the emulator up for Travis CI.

The reason for have a test is that, TensorFlow is transitioning from 1.x to 2.0 and lots of changes happen very frequently. Just want to make sure any other changes introduced will not break the build for Azure storage support.

**Comment by damienpontifex** [comment]
@yongtang I'm still working on the tests. I have them in the subfolder and have been running them with `bazel test //tensorflow_io/azure/...` until now. I'm still to figure out the few changes to make them work in the top level tests folder...and to setup the storage emulator on startup appropriately for the tests to run.
It might be a weekend job at this stage. Thank you for the continued assistance on this :)

**Comment by damienpontifex** [comment]
@yongtang I've been trying to get the tests working. I've now got them running but all failing because it's not connecting properly to the emulator.

I look for an environment variable `TF_AZURE_USE_DEV_STORAGE=1` to setup everything for the https://github.com/azure/azurite storage emulator. I've set this in python, from bash and from the travis.yml. All seem not to be working as I'd seen the error I'm getting of "Unknown failure in container_exists.  No exception, but the container property object is invalid.  errno = 0." before when I didn't have that environment variable set.

I can run the tests with `pytest` locally and everything passes. Any ideas why this might be happening in travis?

**Comment by yongtang** [comment]
@damienpontifex I added the `export TF_AZURE_USE_DEV_STORAGE=1` in `.travis/wheel.test.sh` file (See #327):
https://github.com/tensorflow/io/pull/327/commits/d3358a5435d6d866fb3f6614e8017eddec06e551

That seems to allow test_azure.py runs further with:
```
test_azure.py::AZFSTest::test_also_works_with_full_dns_name PASSED
```

Though the rest of the tests are still failing due to 404 error. Wondering if there are anything else that needs to be updated?

**Comment by damienpontifex** [comment]
🎉 @yongtang thank you. I think I'd put it everywhere except inside that wheel script. Must be something to do with lifetime of the variable maybe, but tests passed! 😀 

Just the couple of failures with kokoro which need resubmitting?

Would you like the commits squashed up before merging for a clean history?

**Comment by yongtang** [comment]
Thanks @damienpontifex 👍 🎉 ! When PR is merged it will be squashed automatically so no need to squash before hand.

The Kokoro CI tests on Linux so there might still be some issues:
1) `.kokorun/io_cpu.sh` needs to invoke `start_azure.sh`, and `start_azure.sh` needs to work on Linux (with `docker run`).
2) There is one failure with read write on Linux:
```
=================================== FAILURES ===================================
________________________ AZFSTest.test_write_read_file _________________________

self = <tests.test_azure.AZFSTest testMethod=test_write_read_file>

    def test_write_read_file(self):
      """Test write/read file.
      """
      # Setup and check preconditions.
      file_name = self._path_to("writereadfile")
      if gfile.Exists(file_name):
        gfile.Remove(file_name)
      # Write data.
      with gfile.Open(file_name, 'w') as w:
        w.write("Hello\n, world!")
      # Read data.
      with gfile.Open(file_name, 'r') as r:
        file_read = r.read()
>       self.assertEqual(file_read, "Hello\n, world!")
E       AssertionError: 'Hello\n, p\x01\x00\x00\x00\x00' != 'Hello\n, world!'

test_azure.py:91: AssertionError
```

**Comment by yongtang** [comment]
@damienpontifex I tried with a commit (9861683) on PR #327 which set up the start_azure.sh on Linux for Kokoro, and skips Linux test's `test_write_read_file`:
```
@pytest.mark.skipif(sys.platform != "darwin", reason=None)
   def test_write_read_file(self):
```

Can you take a look and see if the change could be applied? Once Kokoro passes we should be ready for merge. (We could fix Linux failure of `test_write_read_file` on follow up PRs).

**Comment by yongtang** [comment]
@damienpontifex I think `bash -x -e tests/test_azure/start_azure.sh` also need to be added to `.kokorun/io_cpu.sh`:
https://github.com/tensorflow/io/commit/98616835781cf198e929aef5733042cad813646f#diff-061015047c19d1655d3f0bd75103612f

**Comment by yongtang** [comment]
@damienpontifex All tests passed👍🎊🎉. I will merge the PR for now.

We still need to address the test failure on Linux platform. That could be done in the follow up PR though.

**Comment by damienpontifex** [comment]
Thank you for all the assistance @yongtang 

I'll be much happier now making smaller changes/refinements now we've got this one big one in.

**Pull request merged**